capstone/capdb/utils.py

```diff
@@ -1,36 +1,30 @@
 from django.template.defaultfilters import slugify
 
 
-def generate_unique_slug(model, field, raw_string, count=None):
+def generate_unique_slug(instance, raw_string, field='slug', max_tries=1000):
     """
-    :param field: the field requiring unique slug
-    :param raw_string: basis for slug
-    :return:
-    Rejection sampling: generate unique slug, check for uniqueness
-    if found in model, resample
+        Get a unique slug for instance by sluggifying raw_string, checking database, and appending -1, -2 etc. if necessary.
+        When checking uniqueness, ignore this object itself if it already exists in the database.
+
+        Usage:
+            my_instance.slug = generate_unique_slug(my_instance, 'Obj Title')
     """
 
-    if count:
-        slug = "%s-%s" % (slugify(raw_string[:100]), str(count))
-    else:
-        slug = "%s" % slugify(raw_string[:100])
-        count = 0
+    slug_base = "%s" % slugify(raw_string[:100])
+    for count in range(max_tries):
+        slug = slug_base
+        if count:
+            slug = "%s-%s" % (slug, count)
 
-    kwargs = {field: slug}
-    found = model.objects.filter(**kwargs)
-    if found.count() == 0:
-        return slug
-    else:
-        count += 1
-        return generate_unique_slug(model, field, raw_string, count=count)
+        # ORM query for objects of the same model, with the same slug
+        found = type(instance).objects.filter(**{field: slug})
 
+        # Exclude this exact instance from the query
+        if instance.pk:
+            found = found.exclude(pk=instance.pk)
```

Comment:
Shouldn't this only ever be called when there is no slug?

Reply:
Nope! I take it back. I see that we want to be able to call this whenever, like in the case of jurisdiction slug setting.